### PR TITLE
[MoE Refactor] MXFP4 Cutlass Experts to MK

### PR DESCRIPTION
**Pull request by zyongye**
## Purpose
Refactor MXFP4 cutlass backend for ongoing moe refactor

Also adding testing infrastructure.

## Test Plan
Test GPQA benchmarks, with medium reasoning effort

gpt-oss-120b TP=2 on gb200 with tested kernel on
```bash
VLLM_USE_FLASHINFER_MOE_MXFP4_MXFP8_CUTLASS=1 vllm serve openai/gpt-oss-120b -tp 2
```

gpt-oss-120b TEP=2 on gb200 with tested kernel on
```bash
VLLM_USE_FLASHINFER_MOE_MXFP4_MXFP8_CUTLASS=1 vllm serve openai/gpt-oss-120b -tp 2 -ep
```

gpt-oss-20b TP=2 on H200 with tested kernel on
```bash
VLLM_USE_FLASHINFER_MOE_MXFP4_BF16=1 vllm serve openai/gpt-oss-20b -tp 2
```

Test command: Follow the [recipe](https://docs.vllm.ai/projects/recipes/en/latest/OpenAI/GPT-OSS.html#accuracy-evaluation-panels)

## Test Result
gb200: GPQA with medium reasoning effort on 120b: 0.727. Match the recipe.

H200: GPQA with medium reasoning effort on 20b: 0.6641. Match the recipe.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: .buildkite/test_areas/lm_eval.yaml [hunk 1]
```diff
@@ -73,3 +73,29 @@ steps:
   num_devices: 2
   commands:
     - pytest -s -v evals/gsm8k/test_gsm8k_correctness.py --config-list-file=evals/gsm8k/configs/moe-refactor-dp-ep/config-b200.txt
+
+- label: GPQA Eval (GPT-OSS) (H200)
+  timeout_in_minutes: 120
+  device: h200
```
Comment (robertgshaw2-redhat) [inline .buildkite/test_areas/lm_eval.yaml]: switch to h100 due to resource contraints (we have much more h100 in the ci)

File: vllm/model_executor/layers/fused_moe/config.py [hunk 2]
```diff
@@ -635,6 +660,9 @@ def mxfp4_w4a16_moe_quant_config(
     w2_scale: Union[torch.Tensor, "PrecisionConfig"],
     w1_bias: torch.Tensor | None = None,
     w2_bias: torch.Tensor | None = None,
+    gemm1_alpha: torch.Tensor | None = None,
```
Comment (robertgshaw2-redhat) [inline vllm/model_executor/layers/fused_moe/config.py]: These are just hardcoded values right? (AFAICT they are: https://github.com/zyongye/vllm/blob/d7d68c3127bc27d97b20ceb901068e709c430bd5/vllm/model_executor/layers/quantization/mxfp4.py#L419-L430)

In that case, I think we should avoid passing these via the quant config and instead just having these parameters in the Kernel itself. This will make things clearer and reduce the surface of the API contract

File: vllm/model_executor/layers/quantization/utils/quant_utils.py [hunk 3]
```diff
@@ -151,6 +152,15 @@ def __str__(self):
 kDynamic64Scale = ScaleDesc(torch.float32, False, GroupShape(1, 64))
 kFp8Dynamic64Sym = QuantKey(FP8_DTYPE, kDynamic64Scale, symmetric=True)
 
+kMxfp4DynamicGroupScale = ScaleDesc(MXFP_SCALE_DTYPE, False, GroupShape(1, 32))
```
Comment (robertgshaw2-redhat) [inline vllm/model_executor/layers/quantization/utils/quant_utils.py:158]: Can you add a TODO here to convert this to use scalar type and explain why it was challenging to do in this PR?

File: vllm/model_executor/layers/quantization/mxfp4.py [hunk 4]
```diff
@@ -746,6 +746,30 @@ def _interleave_mxfp4_cutlass_sm90(w):
                 layer.w2_weight_scale = torch.nn.Parameter(
                     w2_scales_interleaved, requires_grad=False
                 )
+
+            assert not self.moe.use_ep, (
```
Comment (robertgshaw2-redhat) [inline vllm/model_executor/layers/quantization/mxfp4.py]: I dont think this is needed. I think this kernel does support EP

NOTE: the noEP thing here is misnamed. It should be NoDPEP

File: vllm/model_executor/layers/fused_moe/utils.py [hunk 5]
```diff
@@ -193,13 +193,10 @@ def _mxfp4_quantize(
 def _mxfp8_e4m3_quantize(
     A: torch.Tensor,
     A_scale: torch.Tensor | None,
-    per_act_token_quant: bool,
-    block_shape: list[int] | None = None,
+    is_sf_swizzled_layout: bool,
```
Comment (robertgshaw2-redhat) [inline vllm/model_executor/layers/fused_moe/utils.py]: I think we should preserve the existing args to just avoid future footguns

File: .buildkite/test_areas/lm_eval.yaml (hunk 1 above)
Comment (mgoin) [reply .buildkite/test_areas/lm_eval.yaml]: ditto, I think H200 is only for 8xH200

File: vllm/model_executor/layers/fused_moe/utils.py (hunk 5 above)
Comment (zyongye) [reply vllm/model_executor/layers/fused_moe/utils.py]: I changed it back. Earlier I thought we should align this with nxfp4 quantization function signature.

File: vllm/model_executor/layers/quantization/mxfp4.py (hunk 4 above)
Comment (zyongye) [reply vllm/model_executor/layers/quantization/mxfp4.py]: The kernel interface actually dispatch to multiple kernels. It will error out when I run EP.

File: .buildkite/test_areas/lm_eval.yaml (hunk 1 above)
Comment (zyongye) [reply .buildkite/test_areas/lm_eval.yaml]: Changed

File: vllm/model_executor/layers/fused_moe/config.py (hunk 2 above)
Comment (zyongye) [reply vllm/model_executor/layers/fused_moe/config.py]: I moved inside the `FlashInferExperts` into the init phase.

File: vllm/model_executor/layers/fused_moe/flashinfer_cutlass_moe.py [hunk 6]
```diff
@@ -119,14 +146,23 @@ def _supports_quant_scheme(
                 ]
                 and p.has_device_capability(90)
             )
-            # fp8 block-scale on 9.0
+            # fp8 block-scale, wmxfp4a16 on 9.0
```
Comment (robertgshaw2-redhat) [inline vllm/model_executor/layers/fused_moe/flashinfer_cutlass_moe.py:155]: is W4A16 supported on B200?

File: vllm/model_executor/layers/fused_moe/flashinfer_cutlass_moe.py [hunk 7]
```diff
@@ -64,10 +67,18 @@ def __init__(
...
+        assert quant_config.weight_quant_dtype in (
+            "mxfp4",
+            "nvfp4",
+            torch.float8_e4m3fn,
+            None,
+        ), (
+            "Only mxfp4, nvfp4, fp8, bfloat16 and"
             " float16 quantization are currently supported."
         )
+        self.device = torch.cuda.current_device()
```
Comment (robertgshaw2-redhat) [inline vllm/model_executor/layers/fused_moe/flashinfer_cutlass_moe.py]: this is already in `moe_config.device`. Its generally not a good idea to call `torch.cuda` if we can avoid it

File: vllm/model_executor/layers/fused_moe/flashinfer_cutlass_moe.py (hunk 6 above)
Comment (zyongye) [reply vllm/model_executor/layers/fused_moe/flashinfer_cutlass_moe.py:155]: Yes. This is running default for gpt-oss on blackwell 
https://github.com/vllm-project/vllm/blob/ab33d2a629be6eca2dd946b1628af4d23d39c547/vllm/model_executor/layers/quantization/mxfp4.py#L138

File: vllm/model_executor/layers/fused_moe/flashinfer_cutlass_moe.py [hunk 8]
```diff
@@ -256,6 +303,44 @@ def apply(
...
+            fc1_expert_biases = self.w1_bias
+            fc2_expert_biases = self.w2_bias
+            swiglu_alpha = self.gemm1_alpha
+            swiglu_beta = self.gemm1_beta
+            swiglu_limit = self.gemm1_clamp_limit
+
+            if self.quant_dtype == "mxfp8":
+                fc1_expert_weights = w1.view(torch.long)
+                fc2_expert_weights = w2.view(torch.long)
+                fake_input_scale = torch.ones(
```
Comment (robertgshaw2-redhat) [inline vllm/model_executor/layers/fused_moe/flashinfer_cutlass_moe.py]: this is being created on the hotpath. we should create this on the cold path (i.e. during init)

File: .buildkite/test_areas/lm_eval.yaml [hunk 9]
```diff
@@ -73,3 +73,29 @@ steps:
...
+  device: h100
+  optional: true
+  num_devices: 2
+  source_file_dependencies:
+  - csrc/
+  - vllm/model_executor/layers/quantization
+  - tests/evals/gpt_oss/
+  commands:
+    - pip install gpt-oss
+    - pytest -s -v evals/gpt_oss/test_gpqa_correctness.py --config-list-file=configs/models-h200.txt
```
Comment (mgoin) [inline .buildkite/test_areas/lm_eval.yaml]: This file name is wrong
```suggestion
    - pytest -s -v evals/gpt_oss/test_gpqa_correctness.py --config-list-file=configs/models-h100.txt
```

File: .buildkite/test_areas/lm_eval.yaml [hunk 10]
```diff
@@ -73,3 +73,29 @@ steps:
...
+  device: b200
+  optional: true
+  num_devices: 2
+  source_file_dependencies:
+  - csrc/
+  - vllm/model_executor/layers/quantization
+  - tests/evals/gpt_oss/
+  commands:
+    - pip install gpt-oss
+    - pytest -s -v evals/gpt_oss/test_gpqa_correctness.py --config-list-file=configs/models-b200.txt
```
Comment (mgoin) [inline .buildkite/test_areas/lm_eval.yaml:101]: If you are going to add these here, please remove the duplicated ones in misc.yaml

File: .buildkite/test_areas/lm_eval.yaml [hunk 11]
```diff
@@ -73,3 +73,29 @@ steps:
...
+  timeout_in_minutes: 120
+  device: b200
+  optional: true
+  num_devices: 2
+  source_file_dependencies:
+  - csrc/
+  - vllm/model_executor/layers/quantization
+  - tests/evals/gpt_oss/
+  commands:
+    - pip install gpt-oss
```
Comment (mgoin) [inline .buildkite/test_areas/lm_eval.yaml]: nit: pin like the original test i.e. `uv pip install --system 'gpt-oss[eval]==0.0.5'`

File: tests/evals/gpt_oss/configs/models-h100.txt [hunk 12]
```diff
@@ -0,0 +1,5 @@
+# H200 model configurations for GPQA evaluation
```
Comment (mgoin) [inline tests/evals/gpt_oss/configs/models-h100.txt]:
```suggestion
# H100 model configurations for GPQA evaluation
```

**Review by mgoin — APPROVED** [review]
LGTM! I kicked off the GPQA Eval tests manually now to see that they work
